File: Go/src/sudoku/sudoku.go
package sudoku

import "fmt"

// Rows are arrays of integers
type Row []int

// A sudoku board is an array of Row
type Sudoku []Row

// Every position in the board has a map associated, saying if a number can be used to fill that position or not
type CandidateNumbers map[int]bool

// Cell represents a position in the Sudoku board
type Cell struct {
	Row              int
	Col              int
	CandidateNumbers CandidateNumbers
}

// Print function
func (s Sudoku) Print() {
	for i, col := range s {
		fmt.Println(i, col)
	}
}

/**
 *	Check if the Sudoku is solved or not. For every row, it does the following:
 * 		- check if there are any zero or if there are repetitions of numbers;
 *		- create the columns, i.e. a row containing elements from the same column position but in different rows;
 *		- create the boxes, i.e. value in different rows and column. Sudoku has 9 boxes, each 3x3;
 *		- for both boxes and columns, repeat the operations done for the row (complete and repetitions)
 */
func (sudoku Sudoku) IsSolved() bool {
	columns := sudoku.getColumns()
	boxes := sudoku.getBoxes()

	for _, Row := range sudoku {

		if !(Row.complete() && Row.repetitions()) {
			return false
		}

	}

	for _, Row := range columns {
		if !(Row.complete() && Row.repetitions()) {
			return false
		}
	}

	for _, Row := range boxes {
		if !(Row.complete() && Row.repetitions()) {
			return false
		}
	}

	return true
}

func (sudoku Sudoku) IsValid() bool {
	columns := sudoku.getColumns()
	boxes := sudoku.getBoxes()

	for _, Row := range sudoku {
		if !(Row.repetitions()) {
			return false
		}
	}

	for _, Row := range columns {
		if !(Row.repetitions()) {
			return false
		}
	}

	for _, Row := range boxes {
		if !(Row.repetitions()) {
			return false
		}
	}

	return true
}

func (sudoku Sudoku) Copy() Sudoku {

	newSudoku := make(Sudoku, 0)

	for _, row := range sudoku {

		newRow := make(Row, 0)

		for _, num := range row {
			newRow = append(newRow, num)
		}

		newSudoku = append(newSudoku, newRow)
	}

	return newSudoku
}

// Update a position in the sudoku board with a value
func (sudoku Sudoku) Update(cell Cell, n int) {

	if n != -1 {
		sudoku[cell.Row][cell.Col] = n
	}

}

// This does the following:
//   - set the candidates number for every cell
//   - get a list of cells who have just one candidate number
//   - fill those cells
func (sudoku Sudoku) MapAndReduce() {

	cells := make([]Cell, 0)

	for rowPos, row := range sudoku {

		for colPos := range row {

			if sudoku[rowPos][colPos] == 0 {

				cell := sudoku.setCandidateNumbers(rowPos, colPos)
				cells = append(cells, cell)

			}

		}

	}

	singleCandidateCells := getSingleCandidateNumberCell(cells)

	for _, cell := range singleCandidateCells {

		n := cell.getSingleNumber()
		sudoku.Update(cell, n)

	}
}

func getSingleCandidateNumberCell(cells []Cell) []Cell {

	singleCells := make([]Cell, 0)

	for _, cell := range cells {

		m := make(map[bool]int)

		for _, value := range cell.CandidateNumbers {
			m[value] = m[value] + 1
		}

		if m[true] == 1 {
			singleCells = append(singleCells, cell)
		}

	}
	return singleCells

}

// Taking a candidate number set of a single cell and starting from a default map,
// it sets to false those numbers which are not proper candidates
func (sudoku Sudoku) setCandidateNumbers(rowPos int, colPos int) Cell {

	boxPos := getBoxPosition(colPos, rowPos)
	candidateMap := defaultMap()
	columns := sudoku.getColumns()
	boxes := sudoku.getBoxes()

	row := sudoku[rowPos]
	column := columns[colPos]
	box := boxes[boxPos]

	for _, n := range row {
		if n > 0 {
			candidateMap[n] = false
		}
	}

	for _, n := range column {
		if n > 0 {
			candidateMap[n] = false
		}
	}

	for _, n := range box {
		if n > 0 {
			candidateMap[n] = false
		}
	}

	return Cell{Row: rowPos, Col: colPos, CandidateNumbers: candidateMap}

}

// Create a default map of candidate numbers, i.e. a map where all numbers are possible candidates
func defaultMap() CandidateNumbers {

	m := make(CandidateNumbers)

	for i := 1; i <= 9; i++ {
		m[i] = true
	}
	return m
}

func (cell Cell) getSingleNumber() int {

	n := -1

	for key, val := range cell.CandidateNumbers {

		if val {
			n = key
		}

	}

	return n

}

// Given the coordinates of a cell in the board, it return the box in which it's contained
func getBoxPosition(col int, row int) int {

	pos := 0

	switch {
	case row < 3:
		switch {
		case col < 3:
			pos = 0
			break
		case col < 6:
			pos = 1
			break
		case col < 9:
			pos = 2
			break
		}
		break
	case row < 6:
		switch {
		case col < 3:
			pos = 3
			break
		case col < 6:
			pos = 4
			break
		case col < 9:
			pos = 5
			break
		}
		break
	case row < 9:
		switch {
		case col < 3:
			pos = 6
			break
		case col < 6:
			pos = 7
			break
		case col < 9:
			pos = 8
			break
		}
		break

	}

	return pos
}

// Check whether a row, column or box has cell that contains zero or not
func (row Row) complete() bool {

	for _, n := range row {

		if n == 0 {
			return false
		}

	}
	return true
}

// Check whether a row, column or box has repetitions of numbers or not
func (row Row) repetitions() bool {

	m := make(map[int]int)

	for _, num := range row {

		if num > 0 {
			m[num] = m[num] + 1
		}

		if m[num] > 1 {
			return false
		}

	}

	return true
}

// Get a map of the columns
func (sudoku Sudoku) getColumns() map[int]Row {

	columns := make(map[int]Row)

	for _, Row := range sudoku {

		for colPos, num := range Row {

			columns[colPos] = append(columns[colPos], num)

		}
	}

	return columns
}

// Get a map of the boxes
func (sudoku Sudoku) getBoxes() map[int]Row {
	boxes := make(map[int]Row)

	for rowPos, Row := range sudoku {

		for colPos, num := range Row {

			boxPos := getBoxPosition(colPos, rowPos)
			boxes[boxPos] = append(boxes[boxPos], num)
		}
	}

	return boxes
}

File: Erlang/src/sudoku/sudoku.erl
-module(sudoku).

%% API
-export([is_solved/1, is_valid/1, init/1, get_candidate/3, get_cell/3, set_cell/4, map_and_reduce/1, print_sudoku/1, search_for/4, to_box/1, set_candidate/3, search_first_empty/1]).


% Print a sudoku board
print_sudoku(Sudoku) ->
  lists:foreach(fun(Row) -> print_row(Row) end, Sudoku).

print_row(Row) ->
  lists:foreach(fun(Cell) -> print_cell(Cell) end, Row),
  io:format("~n").

print_cell(Cell) ->
  case Cell of
    Value when is_integer(Value) ->
      io:format(" ~w", [Value]);
    _ ->
      io:format(" _")
  end.

% Get the value of a cell
get_cell(Sudoku, Row, Col) ->
  V = lists:nth(Row, Sudoku),
  Cell = lists:nth(Col, V),
  ReturnValue = get_cell(Cell),
  {ReturnValue, Row, Col}.

get_cell(Cell) ->
  case Cell of
    Num when is_integer(Num) ->
      Num;
    {'_', _} ->
      '_';
    _ ->
      error({invalid_cell, Cell})
  end.

% Set the value of a cell
set_cell(Sudoku, RowNum, ColNum, Value) ->
  Row = lists:nth(RowNum, Sudoku),
  Cell = lists:nth(ColNum, Row),

  case Cell of
    Num when is_integer(Num) ->  % The cell already has the desired value
      Sudoku;

    {'_', _} ->  % Cell with candidate map
      NewRow = set_nth(ColNum, Row, Value),
      NewSudoku = set_nth(RowNum, Sudoku, NewRow),
      NewSudoku;  % Assuming Value is a valid integer

    '_' ->  % Empty cell
      NewRow = set_nth(ColNum, Row, Value),
      NewSudoku = set_nth(RowNum, Sudoku, NewRow),
      NewSudoku;

    _ ->  % By default
      Sudoku
  end.

set_nth(1, [_|Rest], New) ->
  [New|Rest];
set_nth(I, [E|Rest], New) ->
  [E|set_nth(I-1, Rest, New)].

% Get the list of candidate numbers
get_candidate(Sudoku, RowNum, ColNum) ->
  Row = lists:nth(RowNum, Sudoku),
  Cell = lists:nth(ColNum, Row),
  case Cell of
    {'_', Candidates} ->
      Candidates;
    _ ->
      error({invalid_cell, Cell})
  end.

% Initialization: in every empty list sets the candidate list
init(Sudoku) ->
  %[[init_row(Sudoku, Row) || Row <- Sudoku] || _ <- lists:seq(1, length(Sudoku))].
  lists:map(fun({RowNum, Row}) ->
    init_row(Sudoku, Row, RowNum)
            end, lists:zip(lists:seq(1, length(Sudoku)), Sudoku)).

% Initialization of one row
init_row(Sudoku, Row, RowNum) ->
  lists:map(fun({Col, Item}) -> transform(Sudoku, RowNum, Col, Item) end, lists:zip(lists:seq(1, length(Row)), Row)).


predicate(Cell) ->
  case Cell of
    Num when is_integer(Num) ->
      false;
    '_' ->
      true;
    {'_', _} ->
      true
  end.

% Transform each cell of the form '_' in {'_', candidates}
transform(Sudoku, Row, Col, Cell) ->
  case predicate(Cell) of
    true  -> {'_', set_candidate(Sudoku, Row, Col)};
    false -> Cell
  end.

% Set candidate list for a cell
set_candidate(Sudoku, Row, Col) ->
  Keys = lists:seq(1,9),
  Values = lists:map( fun(V) ->
    not(search_for(Sudoku, Row, Col, V))
    end, Keys),
  CMap = maps:from_list(lists:zip(Keys, Values)),
  maps:filter(fun(_, V) ->
    V =:= true
    end,CMap).

% Return true if a value is already present in the row, column and box of a given cell
search_for(Sudoku, Row, Col, Value) ->
  search_row(Sudoku, Row, Value) or
    search_col(Sudoku, Col, Value) or
    search_box(Sudoku, Row, Col, Value).

% Return true if the value is already in the row
search_row(Sudoku, RowNum, Value) ->
  Row = lists:nth(RowNum, Sudoku),
  lists:member(Value, Row).

% Return true if the value is already in the column
search_col(Sudoku, Col, Value) ->
  Row = lists:nth(Col, to_col(Sudoku)),
  lists:member(Value, Row).

% Return true if the value is already in the box
search_box(Sudoku, Row, Col, Value) ->
  Box = lists:nth(pos(Row,Col), to_box(Sudoku)),
  lists:member(Value, Box).

% Checking if the sudoku board is solved
is_solved(Sudoku) -> complete(Sudoku) and is_valid(Sudoku).

% Check if the sudoku board is valid, i.e. if there are repetitions of numbers
is_valid(Sudoku) -> not(repetitions(Sudoku)).

% Check if there are no zero in the sudoku board
complete(Sudoku) -> complete(Sudoku, 1).

% Given a row, check if there are no zero
complete(Board, Row) when Row =< length(Board) ->
  case check_row(Board, Row, 1) of
    true ->
      false;
    false ->
      complete(Board, Row + 1)
  end;
complete(_, _) ->
  true.

% Check if there are zero in a row
check_row(Sudoku, RowNum, Col) when RowNum =< length(Sudoku) andalso Col =< length(Sudoku) ->
  Row = lists:nth(RowNum, Sudoku),
  case check_cell(Row, Col) of
    Num when is_integer(Num) ->
      check_row(Sudoku, RowNum, Col + 1);
    _ ->
      true
  end;
check_row(_, _, _) ->
  false.

% Given the indexes, get an item from a list
check_cell(Row, Col) -> lists:nth(Col, Row).

% Check if there are repetitions in the rows, columns or boxes
repetitions(Sudoku) -> row_repetitions(Sudoku,1) or col_repetitions(Sudoku, 1) or box_repetitions(Sudoku, 1).

% Check whether there are repetitions in the same box
box_repetitions(Sudoku, Row) ->
  New = to_box(Sudoku),
  row_repetitions(New, Row).

% Get all nine 3x3 boxes in a sudoku board
to_box(Sudoku) ->
  [box(Sudoku, Row, Column) || Row <- [1, 4, 7], Column <- [1, 4, 7]].

box(Sudoku, StartRow, StartColumn) ->
  Rows = lists:sublist(Sudoku, StartRow, 3),
  BoxRows = [lists:sublist(Row, StartColumn, 3) || Row <- Rows],
  Box = lists:flatten(BoxRows),
  Box.

% Check whether there are repetitions in the same column
col_repetitions(Sudoku, Row) ->
  New = to_col(Sudoku),
  row_repetitions(New, Row).

% Transform columns into rows
to_col([]) ->
  [];
to_col([[] | _]) ->
  [];
to_col(Sudoku) ->
  [[lists:nth(N, Row) || Row <- Sudoku] || N <- lists:seq(1, length(Sudoku))].

% Check if there are at least two equal number in the same row
row_repetitions(Sudoku, Row) when Row =< length(Sudoku) ->
  case check_rep_row(Sudoku, Row) of
    true ->
      true;
    false ->
      row_repetitions(Sudoku, Row + 1)
  end;
row_repetitions(_, _) ->
  false.

check_rep_row(Sudoku, Row) ->
  check_rep_row(Sudoku, Row, [0]).

check_rep_row(Sudoku, Row, Seen) when Row =< length(Sudoku) ->
  case check_rep_cell(Sudoku, Row, 1, Seen) of
    true ->
      true;
    false ->
      check_rep_row(Sudoku, Row + 1, Seen)
  end;
check_rep_row(_, _, _) ->
  false.

% Check whether a number was already seen in the list
check_rep_cell(Sudoku, Row, Col, Seen) when Row =< length(Sudoku) andalso Col =< length(Sudoku) ->
  Cell = lists:nth(Col, lists:nth(Row,Sudoku)),
  case Cell of
    '_' ->
      check_rep_cell(Sudoku, Row, Col + 1, Seen); % Empty cell are skipped
    {'_', _} ->
      check_rep_cell(Sudoku, Row, Col + 1, Seen); % Empty cell are skipped
    Num when is_integer(Num) ->
      case lists:member(Num, Seen) of
        true ->
          true; % Repetition found
        false ->
          NewList = lists:append(Seen, [Num]), % Repetition not found and we insert Num in the seen list
          check_rep_cell(Sudoku, Row, Col + 1, NewList)
      end;
    _ ->
      false % Something which isn't a valid cell format
  end;
check_rep_cell(_,_,_,_) ->
  false.

pos(Row, Col) ->
  case Row of
    Num when Num =< 3 ->
      case Col of
        N when N =< 3 ->
          1;
        N when N > 3, N =< 6 ->
          2;
        N when N > 6, N =< 9 ->
          3
      end;
    Num when Num > 3, Num =< 6 ->
      case Col of
        N when N =< 3 ->
          4;
        N when N > 3, N =< 6 ->
          5;
        N when N > 6, N =< 9 ->
          6
      end;
    Num when Num > 6, Num =< 9 ->
      case Col of
        N when N =< 3 ->
          7;
        N when N > 3, N =< 6 ->
          8;
        N when N > 6, N =< 9 ->
          9
      end
  end.

% Find every empty cell which has only one possible candidate and
% fill it with that number
map_and_reduce(Sudoku) ->
  lists:map(fun(Row) ->
    map_and_reduce_row(Row)
    end,Sudoku).

map_and_reduce_row(Row) ->
  lists:map(fun(Elem) ->
    case Elem of
      Num when is_integer(Num) ->
        Num;
      {'_', Candidates} ->
        case single(Candidates) of
          true ->
            N = lists:nth(1, maps:keys(Candidates)),
            N;
            %maps:get(true, Candidates);
          false ->
            {'_', Candidates}
        end
    end
  end, Row).

% Return true if the cell has just one candidate number
single(Candidates) ->
  List = maps:values(Candidates),
  L = length(List),
  if
    L =/= 1 -> false;
    L == 1 -> true
  end.

% Check for the first empty cell occurrence and return its position
search_first_empty(SudokuBoard) ->
  search_first_empty(SudokuBoard, 1).

search_first_empty(Sudoku, Row) when Row =< length(Sudoku) ->
  case search_first_empty(Sudoku, Row, 1) of
    {Row, Col} ->
      {Row, Col};
    false ->
      search_first_empty(Sudoku, Row + 1)
  end;
search_first_empty(_, _) ->
  {0, 0}.

search_first_empty(Sudoku, RowNum, ColNum) when RowNum =< length(Sudoku) andalso ColNum =< length(Sudoku)->
  Cell = lists:nth(ColNum, lists:nth(RowNum, Sudoku)),
  case is_empty(Cell) of
    true ->
      {RowNum, ColNum};
    false ->
      search_first_empty(Sudoku, RowNum, ColNum + 1)
  end;
search_first_empty(_, _, _) ->
  false.

% Check whether a cell is empty or not
is_empty(Cell) ->
  case Cell of
    Num when is_integer(Num) ->
      false;
    {'_', _} ->
      true;
    '_' ->
      true;
    _ ->
      error({invalid_cell, Cell})
  end.